alloc/src/chain.rs

```diff
@@ -0,0 +1,303 @@
+        // SAFETY: active as long as not-null, never give out mut refs.
+        self.ptr.map(|p| unsafe { p.as_ref() })
+    }
+}
+
+/// The node exists inside the allocation owned by `linear`.
+struct ChainNode<A: Allocator> {
+    prev: UnsafeCell<ChainNodePtr<A>>,
+    linear: LinearAllocator<A>,
+}
+
+unsafe impl<A: Allocator + Clone> Send for ChainAllocator<A> {}
+
+impl<A: Allocator> ChainNode<A> {
+    fn remaining_capacity(&self) -> usize {
+        self.linear.remaining_capacity()
+    }
+}
+
+impl<A: Allocator + Clone> ChainAllocator<A> {
+    /// The individual nodes need to be big enough that the overhead of a chain
+    /// is worth it. This is somewhat arbitrarily chosen at the moment.
+    const MIN_NODE_SIZE: usize = 4 * size_of::<ChainNode<A>>();
+
+    /// Creates a new [ChainAllocator]. The `chunk_size_hint` is used as a
+    /// size hint when creating new chunks. Note that there is a minimum size,
+    /// which can change from release to release, and this may not be a
+    /// suitable minimum for your purpose.
+    pub const fn new_in(chunk_size_hint: usize, allocator: A) -> Self {
+        Self {
+            top: UnsafeCell::new(ChainNodePtr::new()),
+            // max is not a const fn, do it manually.
+            node_size: if chunk_size_hint < Self::MIN_NODE_SIZE {
+                Self::MIN_NODE_SIZE
+            } else {
+                chunk_size_hint
+            },
+            allocator,
+        }
+    }
+
+    #[cold]
+    #[inline(never)]
+    fn grow(&self) -> Result<(), AllocError> {
+        let top = self.top.get();
+        let chain_layout = Layout::new::<ChainNode<A>>();
+
+        let linear = {
+            let layout = Layout::from_size_align(self.node_size, chain_layout.align())
+                .map_err(|_| AllocError)?;
+            LinearAllocator::new_in(layout, self.allocator.clone())?
+        };
+
+        // This shouldn't fail.
+        let chain_node_addr = linear
+            .allocate(chain_layout)?
+            .as_ptr()
+            .cast::<ChainNode<A>>();
+        let chain_node = ChainNode {
+            prev: UnsafeCell::new(ChainNodePtr {
+                // SAFETY: todo
+                ptr: unsafe { (*top).ptr },
+            }),
+            linear,
+        };
+
+        // SAFETY: todo
+        unsafe { chain_node_addr.write(chain_node) };
+
+        let chain_node_ptr = ChainNodePtr {
+            // SAFETY: derived from allocation (not null).
+            ptr: Some(unsafe { NonNull::new_unchecked(chain_node_addr) }),
+        };
+        // SAFETY: todo
+        unsafe { self.top.get().write(chain_node_ptr) };
+
+        Ok(())
+    }
+
+    fn capacity_helper(mut ptr: *mut ChainNode<A>) -> usize {
+        let mut capacity = 0_usize;
+        while !ptr.is_null() {
+            capacity += unsafe { (*ptr).linear.capacity() };
+            let prev = unsafe { (*(*ptr).prev.get()).ptr };
+            ptr = unsafe { core::mem::transmute(prev) };
```

Comment:
We don't. I should write a helper which converts `Option<NonNull<T>>` to a `*mut T`. It's a frustrating amount of code for what is a noop at runtime because `Option<NonNull<T>>` and `*mut T` have the same representation. It's something something like:

```rs
    ptr = match prev {
        Some(non_null) => non_null.as_ptr(),
        None => core::ptr::null_mut(),
    };
```

Reply:
I made a helper for now 👍🏻